components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +106,108 @@ impl Dicts {
         }
     }
 
-    fn save_key_dict(&mut self, master_key: &dyn Backend) -> Result<()> {
+    fn open_file_dicts(
+        base: &Path,
+        master_key: &Arc<dyn Backend>,
+    ) -> Result<(Vec<Mutex<FileDictionary>>, usize)> {
+        let mut file_dicts = Vec::with_capacity(FILE_DICT_SIZE);
+        let mut len = 0;
+        for index in 0..FILE_DICT_SIZE {
+            let name = FILE_DICT_NAME.to_owned() + &format!(".{}", index);
+            let file_file = EncryptedFile::new(base, &name);
+            let file_bytes = file_file.read(master_key.as_ref())?;
+            let mut dict = FileDictionary::default();
+            dict.merge_from_bytes(&file_bytes)?;
+            len += dict.files.len();
+            file_dicts.push(Mutex::new(dict));
+        }
+        assert_eq!(file_dicts.len(), FILE_DICT_SIZE);
+        Ok((file_dicts, len))
+    }
+
+    fn save_key_dict(&self, master_key: &dyn Backend, key_dict: &mut KeyDictionary) -> Result<()> {
         let file = EncryptedFile::new(&self.base, KEY_DICT_NAME);
         if !master_key.is_secure() {
-            for value in self.key_dict.keys.values_mut() {
+            for value in key_dict.keys.values_mut() {
                 value.was_exposed = true
             }
         }
-        let key_bytes = self.key_dict.write_to_bytes()?;
+        let key_bytes = key_dict.write_to_bytes()?;
         file.write(&key_bytes, master_key)?;
 
         ENCRYPTION_FILE_SIZE_GAUGE
             .with_label_values(&["key_dictionary"])
             .set(key_bytes.len() as _);
-        ENCRYPTION_DATA_KEY_GAUGE.set(self.key_dict.keys.len() as _);
+        ENCRYPTION_DATA_KEY_GAUGE.set(key_dict.keys.len() as _);
 
         Ok(())
     }
 
-    fn save_file_dict(&self) -> Result<()> {
-        let file = EncryptedFile::new(&self.base, FILE_DICT_NAME);
-        let file_bytes = self.file_dict.write_to_bytes()?;
+    fn save_file_dict(&self, fname: &str) -> Result<()> {
+        let hash = get_index(fname);
+        let name = FILE_DICT_NAME.to_owned() + &format!(".{}", hash);
+        let file = EncryptedFile::new(&self.base, &name);
+        let local_file_dict = self.file_dicts[hash].lock().unwrap();
+        let file_bytes = local_file_dict.write_to_bytes()?;
         // File dict is saved in plaintext.
         file.write(&file_bytes, &PlaintextBackend::default())?;
```

Comment:
we should panic on file write error.

Reply:
How RocksDB handles the returned error?

Reply:
@hicqu fail compaction or flush, which ultimately cause tikv panic. we didn't panic here because I previously want to rely on rocksdb to panic tikv. But since the logic is also used in snapshot, I now think we'd better panic here.

Reply:
So seems for all errors in this method need to cause panic. Is it possible to handle the error in snapshot logic?

Reply:
@hicqu we can. I think we can panic in `DataKeyManager` layer? In the future if key manager get other caller, it is unnecessary for them to know when to panic.

Reply:
or does snapshot logic know better how to panic? like having some cleanup before panic.

Reply:
Snapshot logic can avoid panic in some cases. For example, if it fails when a TiKV receives a snapshot file, the TiKV doesn't need to panic.

And, if we panic here, should all errors in `new_file`, `rename_file` and `link_file` cause panic?

Reply:
okay, let's not change the panic logic in this PR. it is not what this PR want to solve.